fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java

```diff
@@ -743,6 +755,7 @@ private ScoredPlan planIndex(@Nonnull PlanContext planContext, @Nonnull QueryCom
         }
         if (p == null) {
             p = planCandidateScan(candidateScan, indexExpr, filter, sort);
+            p = optimizeCandidateScanForMultidimensionalIndex(planContext, p, index);
```

Comment:
I tried to create the correct plan directly inside of `MultidimensionalAndWithThenPlanner`, however, `AndWithThenPlanner` is used recursively and we are not handing in sufficient information into the recursion (currently) to for instance understand that a comparison is on a dimension as opposed to a prefix column which would require a different logic.

I had a different version of this where there is only one `AndThenPlanner` for value indexes and md indexes  that does not _early out_ but always matches each index expression through all its sub expressions, etc. Then, in a secondary step we can limit value indexes scans to only do prefix scans. That version, however, substantially changed the regular code path and imposed quite a risk of regressing plans.

The example you gave should be planned correctly as the `MultidimensionalAndWithThenPlanner` will match `empty` on `a` and `[(foo]]` on `b`. Then `optimizeCandidateScanFroMultidimensionalIndex` makes sure the match is not breaking the md rules for index matching.

Reply:
What if `indexKeyExpressionForPlan` kept the dimension key and then `planCandidateScan` handled _that_ by recursing with the whole key and then doing what `optimizeCandidateScanForMultidimensionalIndex` does with the result? Would that be clearer? You wouldn't ever have "wrong" intermediate plans.

Reply:
I will think about that, however, I could have some sort of structure that doesn't allow me to disentangle the `KeyExpression` tree into prefix, dimensions, suffix without normalizing it. I think that was my worry when I did it initially.

Wrt to "wrong" intermediate plans. I agree it's not nice, but there is precedence for it, i.e. when the current `AndWithThenPlanner` recurses into the tree it creates a non-functional index scan that is only ever used for book keeping on the way back up.

I would say, this should tell us that it may be nicer to have a recursion to find the matches, i.e. the `ComparisonRanges` and then create a plan when back on top-level, i.e. logically split matching and plan creation. That's how it is done in the Cascades planner which probably gives me some bias but I do think that is generally nice as it doesn't convolute the two steps.

Reply:
So Alec's example without the `AND` is slightly problematic. The index matching will treat this case as a regular index case which may be ok for now. The potential problem occurs if `dim1 < 10` but nothing else. In an unprefixed index that does match. I didn't really want to have completely empty dimensions match (at least not at first so that's how this becomes less of an issue). What is not nice is that if instead the query was just `dim2 < 10` it wouldn't even match as there is no filter on `dim1`. Thus depending on if `dim1` vs `dim2` we would see a difference in behavior which seems not desirable.

Reply:
I refactored the `ScoredPlan` to become a `ScoredMatch` which in the future should allow us to take care of these Alec's problem in a proper way. For now, matching and planning of the match is separated (not intermingled anymore) which takes care of the original comment in a nice way. Unfortunately, in the scope of this PR, we will not change the logic of matching an MD index when not explicitly using an AND filter. That means asymmetries may occur in edge cases. Thus, for now, I will not allow MD indexes to be planned if one or more dimensions are empty-matched.